### PR TITLE
[DSLX] Fixes for casting to/from `xN` (bits constructor) based type, adds docs and more tests

### LINKED ISSUE #1572: Parametric signedness bit type doesn't work
**Describe the bug**
The bytecode emitter errors out if you try to use a return type like `xN[RESULT_SIGNED][RESULT_SZ]`.

```
: INVALID_ARGUMENT: Provided status is not in recognized error form: INTERNAL: Bits-to-array cast bit counts must match. Saw 32 vs 0.; while evaluating: to_int<u32:8, u32:23, u32:32>(APFloat<u32:8, u32:23> { sign: u1:0, bexp: u8:0x0, fraction: u23:0x0 })
...
Error: INTERNAL: Bits-to-array cast bit counts must match. Saw 32 vs 0.; while evaluating: ...
=== Source Location Trace: ===
```

**To Reproduce**
Change this line:

https://github.com/google/xls/blob/69f00f93eb09c0912185ac8a95b8075cb62a7463/xls/dslx/stdlib/apfloat.x#L1301

to `xN[RESULT_SIGNED][RESULT_SZ]` (updating the casts at https://github.com/google/xls/blob/69f00f93eb09c0912185ac8a95b8075cb62a7463/xls/dslx/stdlib/apfloat.x#L1352 and https://github.com/google/xls/blob/69f00f93eb09c0912185ac8a95b8075cb62a7463/xls/dslx/stdlib/apfloat.x#L1355)

**Expected behavior**
This should properly include parameterized signedness and remove the need to cast the result in the wrapper at https://github.com/google/xls/blob/69f00f93eb09c0912185ac8a95b8075cb62a7463/xls/dslx/stdlib/apfloat.x#L1370

### PR DESCRIPTION
**Pull request by cdleary**
Fixes #1572 

Development details -- note that `xN` is the first place where the concept of "BitsLike" became load bearing, any place that assumed `ArrayType` in the hierarchy now has to consider that `ArrayType` could be indicating something that's effectively a bits type; i.e. `BitsLike`. Going forward we always want to use `BitsLike` as the test for a type and whether it's effectively a bits type, there were a few spots here that were missed when the `xN` bits type constructor was introduced.

### CODE REVIEW

File: xls/dslx/stdlib/apfloat.x [hunk 1]
```diff
@@ -1481,14 +1481,15 @@ fn to_signed_or_unsigned_int<RESULT_SZ: u32, RESULT_SIGNED: bool, EXP_SZ: u32, F
         }
     };
 
+    type ReturnT = xN[RESULT_SIGNED][RESULT_SZ];
```
Comment (mikex-oss) [inline xls/dslx/stdlib/apfloat.x:1484]: I think the result is already the correct signedness.

Comment (cdleary) [reply xls/dslx/stdlib/apfloat.x:1484]: I thought this was the change you were trying to make in #1572? Maybe it was just for demonstration purposes -- I can revert this.

Comment (mikex-oss) [reply xls/dslx/stdlib/apfloat.x:1484]: Sorry for the confusion. I was trying to remove the cast on line 1507.

Comment (mikex-oss) [reply xls/dslx/stdlib/apfloat.x]: The RESULT_SIGNED branch now returns sN and the other branch returns uN, which is made possible by the xN change. Thanks btw!

Comment (cdleary) [reply xls/dslx/stdlib/apfloat.x]: No worries! Reverted the apfloat change here. In retrospect, separating that out probably makes sense regardless. I think this'll probably fix the root cause though. 👍

**Comment by hzeller** [comment]
@mikex-oss can you LGTM and tag reviewing internally when ready here ?

**Comment by mikex-oss** [comment]
Hey @richmckeever this looks reasonable to me, can you take a look and approve?

File: xls/dslx/stdlib/apfloat.x (hunk 1 above)
Comment (mikex-oss) [reply xls/dslx/stdlib/apfloat.x]: Thanks, I'll fix up apfloat as a follow up.

File: xls/dslx/bytecode/bytecode_interpreter.cc [hunk 2]
```diff
@@ -602,79 +606,95 @@ absl::Status BytecodeInterpreter::EvalCast(const Bytecode& bytecode,
...
     return absl::InvalidArgumentError(
         "Bytecode interpreter only supports casts from arrays, enums, and "
         "bits; got: " +
-        from.ToString());
+        from_value.ToString());
+  }
+
+  int64_t from_bit_count = from_value.GetBits().value().bit_count();
+
+  // If the thing we're casing from is bits like, and the thing we're casting
```
Comment (richmckeever) [inline xls/dslx/bytecode/bytecode_interpreter.cc:658]: Nit: "casting"

**Review by richmckeever — APPROVED** [review]

File: xls/dslx/bytecode/bytecode_interpreter.cc (hunk 2 above)
Comment (cdleary) [reply xls/dslx/bytecode/bytecode_interpreter.cc]: Done.

File: xls/dslx/stdlib/apfloat.x (hunk 1 above)
Comment (mikex-oss) [reply xls/dslx/stdlib/apfloat.x]: @cdleary where we probably got crossed is that I forgot DSLX doesn't have constexpr if's, so the types in both branches still need to match, while you probably knew that :)

My initial comment was thinking that we could write:

```
    if RESULT_SIGNED {
        // Reduce to the target size, preserving signedness.
        let result = result as sN[RESULT_SZ];
        if !x.sign { result } else { -result }
    } else {
        // Already unsigned, just size correctly.
        result as uN[RESULT_SZ]
    }
```

Anyway, submitted basically what you had with the additional cast in `to_int` removed in https://github.com/google/xls/commit/17d0d09be1f0c895adfd83478e91e728def009b9.

Comment (cdleary) [reply xls/dslx/stdlib/apfloat.x]: You're giving me too much credit, was kind of blindly trying to do what you had described without thinking much. :-) Glad you were able to make the change based off this one though!

**Pull request merged**